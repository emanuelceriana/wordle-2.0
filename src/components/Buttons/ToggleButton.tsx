import styles from "./ToggleButton.module.scss";
import cn from "classnames";

interface ToggleButtonProps {
  disabled?: boolean;
  value: boolean;
  onToggle: () => void;
}

export const ToggleButton = ({
  disabled = false,
  value,
  onToggle,
}: ToggleButtonProps) => {
  return (
    <div className={styles.toggleButton}>
      <label className={styles["toggleButton__switch"]}>
        <input
          data-testid="hardModeToggleButton"
          type="checkbox"
          disabled={disabled}
          checked={value}
          onChange={onToggle}
        />
        <span
          className={cn(styles.slider, styles.round, {
            [styles.disabled]: disabled,
          })}
        ></span>
      </label>
    </div>
  );
};
