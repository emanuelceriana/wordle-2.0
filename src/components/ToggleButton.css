/* The switch - the box around the slider */
.toggleButton .switch {
  position: relative;
  display: inline-block;
  width: 38px;
  height: 20px;
}

/* Hide default HTML checkbox */
.toggleButton .switch input {
  opacity: 0;
  width: 0;
  height: 0;
}

/* The slider */
.toggleButton .slider {
  position: absolute;
  cursor: pointer;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: #ccc;
  -webkit-transition: 0.4s;
  transition: 0.4s;
}

.toggleButton .slider:before {
  position: absolute;
  content: "";
  height: 15px;
  width: 15px;
  left: 4px;
  bottom: 2.8px;
  background-color: white;
  -webkit-transition: 0.4s;
  transition: 0.4s;
}

.toggleButton input:checked + .slider {
  background-color: rgb(40, 175, 40);
}

.toggleButton input:focus + .slider {
  box-shadow: 0 0 1px rgb(65, 226, 65);
}

.toggleButton input:checked + .slider:before {
  -webkit-transform: translateX(15px);
  -ms-transform: translateX(15px);
  transform: translateX(15px);
}

/* Rounded sliders */
.toggleButton .slider.round {
  border-radius: 34px;
}

.toggleButton .slider.round:before {
  border-radius: 50%;
}

.toggleButton .disabled {
  background-color: #ccccccb4;
  cursor: not-allowed;
}

.toggleButton .disabled.slider:before {
  background-color: rgba(201, 198, 198, 0.938);
}
