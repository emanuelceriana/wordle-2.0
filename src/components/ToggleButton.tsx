import "./ToggleButton.css";
import cn from "classnames";

interface ToggleButtonProps {
  disabled: boolean;
}

export const ToggleButton = ({ disabled }: ToggleButtonProps) => {
  return (
    <div className="toggleButton">
      <label className="switch">
        <input type="checkbox" disabled={disabled} />
        <span className={cn("slider", "round", { disabled: disabled })}></span>
      </label>
    </div>
  );
};
