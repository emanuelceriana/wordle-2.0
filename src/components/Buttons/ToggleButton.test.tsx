import { render, fireEvent } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { ToggleButton } from "./ToggleButton";

describe("<ToggleButton />", () => {
  it("should render without errors", () => {
    const { asFragment } = render(
      <ToggleButton value={true} onToggle={() => {}} />,
      {}
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it("should call onToggle when user clicks", () => {
    const value = true;
    const onToggle = jest.fn();
    const { getByTestId } = render(
      <ToggleButton value={value} onToggle={onToggle} />,
      {}
    );

    fireEvent.click(getByTestId("hardModeToggleButton"));

    expect(onToggle).toHaveBeenCalledTimes(1);
  });
  it("should not call onToggle when user clicks if the component is disabled", async () => {
    const value = true;
    const onToggle = jest.fn();

    const { getByTestId } = render(
      <ToggleButton value={value} disabled={true} onToggle={onToggle} />,
      {}
    );

    await userEvent.click(getByTestId("hardModeToggleButton"));

    expect(onToggle).toHaveBeenCalledTimes(0);
  });
});
