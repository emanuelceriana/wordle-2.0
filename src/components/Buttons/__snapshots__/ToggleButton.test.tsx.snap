// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ToggleButton /> should render without errors 1`] = `
<DocumentFragment>
  <div
    class="toggleButton"
  >
    <label
      class="toggleButton__switch"
    >
      <input
        checked=""
        data-testid="hardModeToggleButton"
        type="checkbox"
      />
      <span
        class="slider round"
      />
    </label>
  </div>
</DocumentFragment>
`;
